### PR TITLE
Add decorrelated batch normalization

### PR DESCRIPTION
**Pull request by tkerola**
This PR adds a function and link for decorrelated batch normalization (CVPR 2018): https://arxiv.org/abs/1804.08450

- [x] Forward and backward tests pass.
- [x] Implement group whitening (cf. the paper)
- [x] Add documentation.

This does not implement double backward support nor backward support for `FixedDecorrelatedBatchNormalizationFunction`. I think it's better to leave them unimplemented for now, and add them in another PR.

### CODE REVIEW

**Comment by hvy** [comment]
Is this PR ready to be reviewed, except for the fact that documentations are missing? I agree we can leave parts of the backward/double backward implementations left out and continue with this one.

**Comment by tkerola** [comment]
Yes, I think we can start with review. Sorry, I'll hurry up to finish the documentation and remove the WIP tag.

**Comment by hvy** [comment]
Thank you for the quick action. It is not a hurry (since I'm going through the paper) but I will try to take a look.

File: chainer/functions/normalization/decorrelated_batch_normalization.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+
+
+class DecorrelatedBatchNormalizationFunction(function_node.FunctionNode):
+
+    def __init__(self, groups=16, eps=2e-5, mean=None, projection=None,
+                 decay=0.9):
+        self.groups = groups
+
+        self.expected_mean = mean
+        self.expected_projection = projection
```
Comment (hvy) [inline chainer/functions/normalization/decorrelated_batch_normalization.py]: How about `running_{mean,projection}` to align with BN?

File: chainer/functions/normalization/decorrelated_batch_normalization.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        C = c // g
+        m = b * g
+        for i in spatial_axis:
+            m *= x.shape[i]
+
+        if self.expected_mean is None:
+            self.expected_mean = xp.zeros(C, dtype=x.dtype)
+            self.expected_projection = xp.eye(C, dtype=x.dtype)
+
+        if self.groups > 1:
```
Comment (hvy) [inline chainer/functions/normalization/decorrelated_batch_normalization.py]: It's possible to reuse `g` here I think.

File: chainer/functions/normalization/decorrelated_batch_normalization.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+    def backward(self, indexes, grad_outputs):
+        gy, = grad_outputs
+
+        f = DecorrelatedBatchNormalizationGrad(
+            self.groups, self.eigvals, self.eigvectors, self.y_hat_pca)
+        return f(gy),
+
+
+class DecorrelatedBatchNormalizationGrad(function.Function):
```
Comment (hvy) [inline chainer/functions/normalization/decorrelated_batch_normalization.py]: Let's use the new `FunctionNode` (only need to change the above `f(gy),` to `f.apply((gy,))`.

File: tests/chainer_tests/functions_tests/normalization_tests/test_decorrelated_batch_normalization.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+    'groups': [1, 2],
+    'eps': [2e-5, 5e-1],
+    'dtype': [numpy.float32],
+    'c_contiguous': [True, False],
+}) + testing.product({
+    'n_channels': [8],
+    'ndim': [1],
+    'groups': [1, 2],
+    'eps': [2e-5, 5e-1],
+    # NOTE(tommi): xp.linalg.eigh does not support float16
```
Comment (hvy) [inline tests/chainer_tests/functions_tests/normalization_tests/test_decorrelated_batch_normalization.py]: I think this comment applies to `np` but not `cp`?

File: tests/chainer_tests/links_tests/normalization_tests/test_decorrelated_batch_normalization.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        gradient_check.check_backward(
+            self.link, x_data, y_grad, (),
+            eps=1e-2, **self.check_backward_options)
+
+    @condition.retry(3)
+    def test_backward_cpu(self):
+        self.check_backward(self.x, self.gy)
+
+    @attr.gpu
+    @condition.retry(3)
```
Comment (hvy) [inline tests/chainer_tests/links_tests/normalization_tests/test_decorrelated_batch_normalization.py]: Is it possible to adjust tolerances to avoid relying on `retry` ? Preferably (we do this internally), you'd like to run the test for a couple of thousand repetitions without a single failure.

File: chainer/links/normalization/decorrelated_batch_normalization.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    In testing mode, it uses pre-computed population statistics to normalize
+    the input variable. The population statistics is approximated if it is
+    computed by training mode, or accurate if it is correctly computed by
+    fine-tuning mode.
+
+    Args:
+        size (int or tuple of ints): Size (or shape) of channel
+            dimensions.
+        groups (int): Number of groups to use for group whitening.
+        decay (float): Decay rate of moving average. It is used on training.
```
Comment (hvy) [inline chainer/links/normalization/decorrelated_batch_normalization.py]: How about `Decay rate of moving average which is used during training.` ?

File: chainer/links/normalization/decorrelated_batch_normalization.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+       :func:`~chainer.functions.decorrelated_batch_normalization`,
+       :func:`~chainer.functions.fixed_decorrelated_batch_normalization`
+
+    Attributes:
+        expected_mean (numpy.ndarray or cupy.ndarray): Population mean.
+        expected_projection (numpy.ndarray or cupy.ndarray): Population
+            projection.
+        groups (int): Number of groups to use for group whitening.
+        N (int): Count of batches given for fine-tuning.
+        decay (float): Decay rate of moving average. It is used on training.
```
Comment (hvy) [inline chainer/links/normalization/decorrelated_batch_normalization.py]: Ditto.

**Comment by hvy** [comment]
@tkerola I added some initial comments. Could you take a look?

**Comment by hvy** [comment]
For the record, discussed with @tkerola and agreed to add `gamma` and `beta` to DBN too.

File: chainer/functions/normalization/decorrelated_batch_normalization.py [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+        gx = gx_hat.reshape((C, b * g,) + gy.shape[2:]).transpose(
+            (1, 0) + spatial_axis)
+        if self.groups > 1:
+            gx = gx.reshape((-1, c, ) + gy.shape[2:])
+
+        self.retain_outputs(())
+        return gx,
+
+    def backward(self, inputs, grad_outputs):
+        raise NotImplementedError
```
Comment (hvy) [inline chainer/functions/normalization/decorrelated_batch_normalization.py]: Could you add short TODO comment, preferably explaining why it's not yet implemented?

File: chainer/functions/normalization/decorrelated_batch_normalization.py [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+            y = y.reshape((-1, c, ) + x.shape[2:])
+
+        return y,
+
+    def backward(self, indexes, grad_outputs):
+        raise NotImplementedError
+
+
+def _sym(x):
+    return (x + x.T) / 2
```
Comment (hvy) [inline chainer/functions/normalization/decorrelated_batch_normalization.py]: It seems like `_sym` is only called once. Should we inline it?

**Comment by hvy** [comment]
Let me bump this one.

**Comment by tkerola** [comment]
@toslunar Has agreed to kindly take over this PR, since I don't have much time currently to complete it. :bowing_man:

**Comment by hvy** [comment]
Let me close this PR as the one taking over it has been merged.
